### PR TITLE
Refactor: new requirements for `VestedRewardsDistributionJob`

### PR DESCRIPTION
**Pull request by amusingaxl**
This PR addresses the need to cope with calls to `VestedRewardsDistributionLike.distribute()` calls that might revert.

It also addresses the improvements described in #31.

### CODE REVIEW

File: src/VestedRewardsDistributionJob.sol
Comment (0xp3th1um) [inline src/VestedRewardsDistributionJob.sol:140]: This packs two functions in one in a nice way, but it assumes that if `dist` does not exist then it is a valid one and there is no mistake in the address provided. The two-step process would be less error prone but I think it is fine since they are not added manually.

File: src/VestedRewardsDistributionJob.sol
Comment (0xp3th1um) [inline src/VestedRewardsDistributionJob.sol:218]: This works but it is too confusing. We keep both `workable()` and `work()` to not break the interface, but is seems to me that `work()` will rarely be used. Am i right?

File: src/tests/VestedRewardsDistribution-integration.t.sol
Comment (0xp3th1um) [inline src/tests/VestedRewardsDistribution-integration.t.sol:135]: I think since `work()` is no longer a view function it should be also tested to prove that it works as expected i.e. it calls `work()` and it has the desired side effects.

**Review by 0xp3th1um — COMMENTED** [review]
Looks good, i just left some comments before i approve

File: src/VestedRewardsDistributionJob.sol
Comment (amusingaxl) [reply src/VestedRewardsDistributionJob.sol:140]: The objective here is to reduce the API size because this is going to be managed through spells.

File: src/VestedRewardsDistributionJob.sol
Comment (amusingaxl) [reply src/VestedRewardsDistributionJob.sol:218]: Both are used the same way as before:
- The keeper will call `workable` using something like [`ethers.js#staticCall`](https://docs.ethers.org/v6/api/contract/#BaseContractMethod-staticCall), which is pretty much like a transaction simulation, however the transaction is not actually submitted, so the state is not updated.
- If the `staticCall` returns `(true, <address>)`, then the keeper will call `work` with the proper arguments.

It's true that if the keeper `staticCall`s `workable()` to check an action is required, and then submits a transaction calling `workable()` again, it would have the same effect, however `workable()` is designed to look for work to be done and therefore is much more gas intensive.

File: src/tests/VestedRewardsDistribution-integration.t.sol
Comment (amusingaxl) [reply src/tests/VestedRewardsDistribution-integration.t.sol:135]: Do you mean `workable`?

Comment (0xp3th1um) [reply src/tests/VestedRewardsDistribution-integration.t.sol:135]: I meant `workable()`!

File: src/VestedRewardsDistributionJob.sol
Comment (0xp3th1um) [reply src/VestedRewardsDistributionJob.sol:218]: I see, so what you say is that there is no incentive to just  call `workable()` i.e. send an actual tx and waste gas...

File: src/tests/VestedRewardsDistribution-integration.t.sol
Comment (amusingaxl) [reply src/tests/VestedRewardsDistribution-integration.t.sol:135]: If you look closely, there are some assertions through `test_work` and others that test the output of `workable`.
I tried to split them into separate test cases, but it just gets more verbose for the same effect.

**Review by 0xp3th1um — APPROVED** [review]
good job!

File: src/tests/VestedRewardsDistribution-integration.t.sol
Comment (oddaf) [inline src/tests/VestedRewardsDistribution-integration.t.sol]: Console not in use in the contract.

File: src/VestedRewardsDistributionJob.sol
Comment (oddaf) [inline src/VestedRewardsDistributionJob.sol:218]: Just something to maybe note to the keeper network operators: ETH_CALL does have a gas limit providers use to prevent DoS. The limits are very high though (500M gas on Alchemy, 10x the current mainnet block gas limit on Infura, couldn't find the limit on Tenderly).

This function could revert if there are enough workable jobs (as it will work through all of them). With current amount of jobs this is highly unlikely (~220k gas per job). If they rely on their own nodes they should review what limit they are enforcing (if any).

File: src/tests/VestedRewardsDistribution-integration.t.sol
Comment (amusingaxl) [reply src/tests/VestedRewardsDistribution-integration.t.sol]: Oops!

File: src/VestedRewardsDistributionJob.sol
Comment (amusingaxl) [reply src/VestedRewardsDistributionJob.sol:218]: Thanks for the info, I was actually unaware of that. I'll add a note to NatSpec regarding this limitation.

> (~220k gas per job)
I wonder how did you get to this figure. Runnig `forge test --gas-report` gives me:
image

In any case, this seems like a highly unlikely scenario, it would require hundreds or maybe thousands of jobs to reach the existing limits.

**Review by oddaf — APPROVED** [review]

**Pull request merged**
